### PR TITLE
Add delayed deletion of rowsets function, fix -230 error.

### PR DESCRIPTION
**Pull request by ZhangYu0123**
Related issue #4017, main changes as follows:
1. Add expired_snapshot_rs_version_map，_expired_snapshot_rs_metas，
2. Add  VersionedRowsetTracker record compacted path version
3. Record path version when rowsets compact
4. In gc process, add expired snapshot rowsets to unused set to remove.

### CODE REVIEW

File: be/src/common/config.h [hunk 1]
```diff
@@ -230,6 +230,8 @@ namespace config {
     CONF_mInt32(pending_data_expire_time_sec, "1800");
     // inc_rowset expired interval
     CONF_mInt32(inc_rowset_expired_sec, "1800");
+    // inc_rowset snapshot rs sweep time interval
+    CONF_mInt32(tablet_rowset_expired_snapshot_sweep_time, "1800");
```
Comment (wuyunfeng) [inline be/src/common/config.h:234]: It is better to add `unit` to your config item

File: be/src/olap/rowset_graph.cpp [hunk 2]
```diff
@@ -17,14 +17,247 @@
...
+
+    auto iter = other_path.begin();
+    for (; iter != other_path.end(); iter++) {
+        Edges edges = iter->second;
+        auto min_begin_version = iter->first;
+
+        while (!edges->empty()) {
+            PathVersionListSharedPtr path_version_ptr(new std::vector<VersionTrackerSharedPtr>());
+            // 1. find a path
+            //   begin from min_begin_version
```
Comment (wuyunfeng) [inline be/src/olap/rowset_graph.cpp:64]: merge this two comment into one line maybe better

File: be/src/olap/rowset_graph.cpp [hunk 3]
```diff
@@ -17,14 +17,247 @@
...
+            auto min_end_version = edges->begin()->first;
+            auto min_rs_meta = edges->begin()->second;
+            //   tracker the first
```
Comment (wuyunfeng) [inline be/src/olap/rowset_graph.cpp:67]:
```suggestion
            // tracker the first
```

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -61,7 +61,10 @@ Tablet::Tablet(TabletMetaSharedPtr tablet_meta, DataDir* data_dir) :
         _last_cumu_compaction_success_millis(0),
         _last_base_compaction_success_millis(0),
         _cumulative_point(kInvalidCumulativePoint) {
-    _rs_graph.construct_rowset_graph(_tablet_meta->all_rs_metas());
+    // change _rs_graph to _versioned_rs_tracker
+    // _rs_graph.construct_rowset_graph(_tablet_meta->all_rs_metas());
+    _versioned_rs_tracker.construct_versioned_tracker(_tablet_meta->all_rs_metas(),
```
Comment (wuyunfeng) [inline be/src/olap/tablet.cpp:66]: 66 - 67 has nonstandard format

**Review by wuyunfeng — CHANGES_REQUESTED** [review]

File: be/src/olap/tablet.cpp [hunk 5]
```diff
@@ -310,7 +321,10 @@ OLAPStatus Tablet::add_inc_rowset(const RowsetSharedPtr& rowset) {
     RETURN_NOT_OK(_tablet_meta->add_rs_meta(rowset->rowset_meta()));
     _rs_version_map[rowset->version()] = rowset;
     _inc_rs_version_map[rowset->version()] = rowset;
-    _rs_graph.add_version_to_graph(rowset->version());
+
+    // _rs_graph.add_version_to_graph(rowset->version());
```
Comment (wutiangan) [inline be/src/olap/tablet.cpp:325]: remove this line？

File: be/src/olap/tablet.cpp [hunk 6]
```diff
@@ -358,9 +383,102 @@ void Tablet::delete_expired_inc_rowsets() {
...
+        Version test_version = Version(first_version->version().first, end_version->version().second);
+
+        OLAPStatus status = capture_consistent_versions(test_version, nullptr);
+        // When there is no consistent versions, we must reconstruct the tracker.
+        if (status != OLAP_SUCCESS) {
+            LOG(WARNING) << "The consistent version check fails, there are bugs. "
+                << "Reconstruct the tracker to recover versions in tablet=" << tablet_id();
+
+            _timestamped_version_tracker.reconstruct_versioned_tracker(
+                    _tablet_meta->all_rs_metas(), expired_snapshot_version_path_map);
```
Comment (morningman) [inline be/src/olap/tablet.cpp:422]: expired_snapshot_version_path_map does not contains the expired path you just deleted.

File: be/src/olap/tablet.cpp [hunk 7]
```diff
@@ -358,9 +383,102 @@ void Tablet::delete_expired_inc_rowsets() {
...
+                // delete rowset
+                StorageEngine::instance()->add_unused_rowset(it->second);
+                _expired_snapshot_rs_version_map.erase(it);
+                LOG(INFO) << "delete expired snapshot rowset tablet=" << full_name() 
+                    <<" version[" << timestampedVersion->version().first << "," << timestampedVersion->version().second 
+                    << "] move to unused_rowset success " << std::fixed << expired_snapshot_sweep_endtime;
+            } else {
+                LOG(WARNING) << "delete expired snapshot rowset tablet=" << full_name() 
+                    <<" version[" << timestampedVersion->version().first << "," << timestampedVersion->version().second 
+                    << "] not find in expired snapshot rs version ";
```
Comment (morningman) [inline be/src/olap/tablet.cpp:460]:
```suggestion
                    << "] not find in expired snapshot rs version map";
```

File: be/src/olap/tablet.cpp [hunk 8]
```diff
@@ -358,9 +383,102 @@ void Tablet::delete_expired_inc_rowsets() {
...
+            }
+            _delete_expired_snapshot_rowset_by_version(timestampedVersion->version());
```
Comment (morningman) [inline be/src/olap/tablet.cpp:462]: in `_delete_expired_snapshot_rowset_by_version`, we call `_expired_snapshot_rs_version_map.erase()` again, which is duplicated.

File: be/src/olap/tablet_meta.cpp [hunk 9]
```diff
@@ -544,6 +548,27 @@ OLAPStatus TabletMeta::add_inc_rs_meta(const RowsetMetaSharedPtr& rs_meta) {
     return OLAP_SUCCESS;
 }
 
+void TabletMeta::delete_expired_snapshot_rs_meta_by_version(const Version& version) {
+    auto it = _expired_snapshot_rs_metas.begin();
+    while (it != _expired_snapshot_rs_metas.end()) {
+        if ((*it)->version() == version) {
+            _expired_snapshot_rs_metas.erase(it);
+            break;
```
Comment (morningman) [inline be/src/olap/tablet_meta.cpp:556]: Are you sure there will be only one version which equals to the specified version?

**Review by morningman — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**